Remove runtime type-checking from enumerable types

aisamanra opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->


### Motivation
This removes runtime type-checking for collections from Sorbet.

This type-checking is a frequent source of slowness, especially in surprising situations, and as per offline conversation we've decided to remove it.


### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

aisamanra commented:
One question I have: is it worth capturing the existing behavior of runtime type-checking for an enumerable in a helper method somewhere? I'm thinking that someone might want to have something like

```ruby
sig {params(xs: T::Array[Integer]).void}
def foo(xs)
  T.deep_type_assert!(xs, T::Array[Integer])
  # foo
end
```

which does the same thing as the current runtime verification, if people want to opt-in to that extra type-checking in some situations, but that it won't happen by default.

aisamanra commented:
(Ah, I was lulled into a false sense of security by just seeing the LSP test failing in CI—I thought that meant it was the only one.)

I've just removed all those tests, since they were testing invariants we no longer expect to hold after this.

N.B. we should do internal comms when this lands, since it'll change expected runtime behavior.

jez commented on gems/sorbet-runtime/test/types/props/decorator.rb:
```diff
@@ -172,18 +172,6 @@ class InnerStruct < T::Struct
     prop :the_hash, T::Hash[String, InnerStruct]
   end
 
-  describe 'validating prop values' do
```
Instead of deleting all these tests, can we convert them into tests that capture the new behavior?

bbuchalter commented:
I recently uncovered this problem as well. Here's some performance metrics https://github.com/bbuchalter/sorbet_sig_performance

tl;dr 
When large hashes are fetched and type-checked at runtime, Sorbet has to run checks on every member in hash- which is more expensive than expected.
We have observed a linear degradation in performance when comparing the time it takes to read a hash with a sig vs. a hash with `.checked(:never)` ([doesn't run runtime checks](https://sorbet.org/docs/runtime#checked-whether-to-check-in-the-first-place)) specified.
For a `Hash[String, String]`:

| # of keys  |  Avg Method Calls/5 seconds (with sig) | Avg Method calls/5 seconds (sig never checked)  | Order of Magnitude of diff  |   |
|---|---|---|---|---|
| 5  | 454,777  |  2,308,000 | 1|
|  50 | 643,000  | 86,662,000  |  2| 
|  5000 | 7,020  |  96,780,000 | 4|

jez commented on gems/sorbet-runtime/test/types/types.rb:
```diff
@@ -212,13 +212,10 @@ def subtype_of?(other)
-        expected_error = "Expected type T::Array[Integer], " \
-                         "got T::Array[T::Boolean]"
-        assert_equal(expected_error, msg)
+        assert_nil(type.error_message_for_obj(value))
```
@aisamanra What does it mean for `error_message_for_obj` to be `nil`? Do we still show an "expected / actual" error message?

jez replied:
Oh I see, these are all saying that `error_message_for_obj` is `nil` because there is no error raised, because the only problem is that the element type is mismatched, not the class type.

jez commented on gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:
```diff
@@ -23,50 +23,7 @@ def name
-      when Enumerator
-        # Enumerators can be unbounded: see `[:foo, :bar].cycle`
-        return true
-      when Range
-        # A nil beginning or a nil end does not provide any type information. That is, nil in a range represents
-        # boundlessness, it does not express a type. For example `(nil...nil)` is not a T::Range[NilClass], its a range
-        # of unknown types (T::Range[T.untyped]).
-        # Similarly, `(nil...1)` is not a `T::Range[T.nilable(Integer)]`, it's a boundless range of Integer.
-        (obj.begin.nil? || @type.valid?(obj.begin)) && (obj.end.nil? || @type.valid?(obj.end))
```
Maybe it doesn't make sense to do this, but `Range` always checks a constant number of elements (`.begin` and `.end`). It might be easy enough to keep this behavior, but I'm also fine removing it as you have here.

jez commented:
@aisamanra It sounds like you were going to do some communications on Slack to announce this change.

Would you also mind updating the (fragment of a) page at https://sorbet.org/docs/stdlib-generics to mention maybe in a sentence or two that the type argument is only a hint to the static system, and is never checked at runtime (and maybe also why / the history?)

aisamanra commented:
I went ahead and added the docs to in this PR. (I also added an extra sentence as long as I was already modifying those docs about the use of bare `Array` and the like: I was already there, after all!)

I'll also draft an internal email really quickly to do comms about this; we can discuss that offline.
